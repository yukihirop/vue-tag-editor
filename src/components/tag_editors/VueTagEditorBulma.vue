<template>
  <span>
    <tags
      :tags="tags"
      :type="type"
      :event-hub="eventHub"
      :tagAreaClass="tagAreaClass"
      :tagContentClass="tagContentClass"
      :deleteAreaClass="deleteAreaClass"
      :deleteContentClass="deleteContentClass"
      :backgroundColorClass="backgroundColorClass"
    />
    <input
      :class="inputContentClass"
      v-model="tag"
      placeholder="Add tags..."
      @keyup.enter="inputTagWithEmit"
    >
  </span>
</template>

<script>
import Vue from 'vue'
import Tags from '../Tags.vue'

export default {
  name: 'VueTagEditorBulma',
  components:{
    Tags: Tags
  },
  props: {
    tags:{
      type: Array,
      default(){
        return []
      }
    },
    type: {
      type: String,
      default: 'label'
    },
    backgroundColorClass: {
      type: String,
      default: ""
    },
    inputContentClass: {
      type: String,
      default: ""
    }
  },
  data(){
    return {
      tag: '',
      isAddTag: false,
      eventHub: new Vue(),
      tagAreaClass: "tagArea",
      tagContentClass: "tagContent",
      deleteAreaClass: "deleteArea",
      deleteContentClass: "deleteContent"
    }
  },
  mounted(){
    this.eventHub.$on('click-tag', this._emitClickTag)
    this.eventHub.$on('delete-tag', this._emitDeleteTag)
  },
  methods: {
    inputTagWithEmit(){
      var tag = this.tag
      if (this._enableAdd(this.tag)) {
        this.tags.push(this.tag)
        this.isAddTag = true
      }
      this.tag = null

      this.$nextTick(() => {
        this.$emit('handler-after-input-tag', tag, this.isAddTag)
        this.isAddTag = false
      })
    },
    _enableAdd(tag){
      return (this.tags.indexOf(tag) == -1) && tag != undefined || ''
    },
    _emitClickTag(tag){
      this.$emit('handler-after-click-tag', tag)
    },
    _emitDeleteTag(tag){
      this.$emit('handler-after-delete-tag', tag)
    }
  }
}
</script>

<style lang="scss" scoped>
@import "../../assets/sass/bulma.scss";

/deep/ .tagArea {
  @extend .tag:not(body);
  @extend .is-primary;
  padding: 2px 4px 2px 4px;
  margin: 2px 4px 2px 0px;
}
/deep/ .tagContent {
}
/deep/ .deleteArea {
  @extend .is-primary;
  background-color: transparent;
  border: none;
}
/deep/ .deleteContent {
  @extend .is-primary;
  color: white;
}
</style>>