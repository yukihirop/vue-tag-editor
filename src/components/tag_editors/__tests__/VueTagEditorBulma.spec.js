import { createLocalVue, shallowMount } from '@vue/test-utils'
import VueTagEditorBulma from '../VueTagEditorBulma'

const localVue = createLocalVue()

describe('VueTagEditorBulma', () => {
  const wrapper = shallowMount(VueTagEditorBulma, {
    localVue,
    propsData: {
      tags: ['javascript', 'ruby']
    }
  })

  it('correct name options', () => {
    expect(wrapper.vm.$options.name).toBe('VueTagEditorBulma')
  })

  it('correct default props', () => {
    expect(wrapper.props().type).toBe('label')
  })

  it('renders correctly (snapshots)', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('render the correct markup', () => {
    expect(wrapper.html()).toContain("tags=\"javascript,ruby\" type=\"label\"")
    expect(wrapper.html()).toContain("Add tags...")
  })

  it('has a input', () => {
    expect(wrapper.contains('input')).toBe(true)
  })

  describe('Cursor up create tag', () => {
    const input = wrapper.find('input')

    // Can not test update props
    // https://github.com/vuejs/vue-test-utils/issues/480#issuecomment-374744653
    describe('input tag do not exists', () => {
      it('create tag', (done) => {
        expect.assertions(3)

        input.setValue('swift')
        wrapper.vm.inputTagWithEmit()

        wrapper.vm.$nextTick(() => {
          expect(wrapper.emitted('handler-after-input-tag')).toBeTruthy()
          expect(wrapper.emitted('handler-after-input-tag').length).toEqual(1)
          expect(wrapper.emitted('handler-after-input-tag')[0]).toEqual(['swift', true])
          done()
        })
      })
    })
  })

  describe('should emitted after click tag', () => {
    wrapper.vm._emitClickTag('test')
    expect(wrapper.emitted('handler-after-click-tag')).toBeTruthy()
    expect(wrapper.emitted('handler-after-click-tag').length).toEqual(1)
    expect(wrapper.emitted('handler-after-click-tag')[0]).toEqual(['test'])
  })
})