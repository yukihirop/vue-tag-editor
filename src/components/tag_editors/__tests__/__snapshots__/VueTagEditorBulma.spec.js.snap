// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VueTagEditorBulma renders correctly (snapshots) 1`] = `
<span>
  <tags-stub
    deleteareaclass="deleteArea"
    deletecontentclass="deleteContent"
    eventhub="[object Object]"
    tagareaclass="tagArea"
    tagcontentclass="tagContent"
    tagcustomclass=""
    tags="javascript,ruby"
    type="label"
  />
   
  <input
    class=""
    placeholder="Add tags..."
  />
</span>
`;
